### PR TITLE
Create new M.IM.T.Extensions library to provide crypto operations using cloud-based security keys.

### PR DESCRIPTION
**Pull request by CBaud**
The current library provides an extensible model for custom signing and encryption solutions. This makes it possible to create tokens that are backed by cloud-based security keys, such as Azure Key Vault. These changes address #936 and #980 by creating a new extensions project that provides the base implementation to create those types of tokens.

### CODE REVIEW

File: buildConfiguration.xml [hunk 1]
```diff
@@ -15,6 +15,7 @@
       <project name="Microsoft.IdentityModel.Protocols" />
       <project name="Microsoft.IdentityModel.Protocols.WsFederation" />
       <project name="Microsoft.IdentityModel.Protocols.OpenIdConnect" />
+	   <project name="Microsoft.IdentityModel.Tokens.Extensions" />
     </src>
```
Comment (mafurman) [inline buildConfiguration.xml]: Looks like the spacing here is a bit off...this is something I must have overlooked when I added this file in.

File: test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultKeyWrapProviderTests.cs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+            _client = new MockKeyVaultClient();
+            _key = new KeyVaultSecurityKey(KeyVaultUtilities.CreateKeyIdentifier(), keySize: default);
+        }
+
+        public static TheoryData<KeyWrapProviderTheoryData> DisposeProviderTheoryData
+        {
+            get => new TheoryData<KeyWrapProviderTheoryData>
+            {
+                new KeyWrapProviderTheoryData
+                {
```
Comment (mafurman) [inline test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultKeyWrapProviderTests.cs]: Maybe put "First = true" here just to be consistent?

File: test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultKeyWrapProviderTests.cs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                    Algorithm = SecurityAlgorithms.RsaOAEP,
+                    ExpectedException = ExpectedException.NoExceptionExpected,
+                    TestId = nameof(SecurityAlgorithms.RsaOAEP),
+                },
+            };
+        }
+
+        [Theory, MemberData(nameof(DisposeProviderTheoryData))]
```
Comment (mafurman) [inline test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultKeyWrapProviderTests.cs]: I believe in most of our tests we put the TheoryData directly after the test method that is using it.

File: test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultSignatureProviderTests.cs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            _key = new KeyVaultSecurityKey(KeyVaultUtilities.CreateKeyIdentifier(), keySize: default);
+        }
+
+        public static TheoryData<SignatureProviderTheoryData> DisposeProviderTheoryData
+        {
+            get => new TheoryData<SignatureProviderTheoryData>
+            {
+                new SignatureProviderTheoryData
+                {
+                    Algorithm = SecurityAlgorithms.RsaSha256,
```
Comment (mafurman) [inline test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultSignatureProviderTests.cs]: First = true?

File: buildConfiguration.xml (hunk 1 above)
Comment (CBaud) [reply buildConfiguration.xml]: Fixed: replaced a tab with whitespace.

File: test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultKeyWrapProviderTests.cs (hunk 2 above)
Comment (CBaud) [reply test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultKeyWrapProviderTests.cs]: Fixed: Added "First = true"

File: test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultKeyWrapProviderTests.cs (hunk 3 above)
Comment (CBaud) [reply test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultKeyWrapProviderTests.cs]: Fixed: Moved the TheoryData directly after the test methods that are using them.

File: test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultSignatureProviderTests.cs (hunk 4 above)
Comment (CBaud) [reply test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultSignatureProviderTests.cs]: Fixed.

File: test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultSecurityKeyConfidentialClientTheoryData.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+//
+//------------------------------------------------------------------------------
+
+using System;
+
+namespace Microsoft.IdentityModel.Tokens.Extensions.Tests
+{
+    public class KeyVaultSecurityKeyConfidentialClientTheoryData : KeyVaultSecurityKeyTheoryData
+    {
+        public string ClientId { get; set; } = $"{Guid.NewGuid():D}";
```
Comment (GeoK) [inline test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultSecurityKeyConfidentialClientTheoryData.cs]: Just curios, any difference between ` $"{Guid.NewGuid():D}" ` and `Guid.NewGuid().ToString()`?

File: test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultKeyWrapProviderTests.cs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+                var keyBytes = Guid.NewGuid().ToByteArray();
+                var wrappedKey = provider.WrapKey(keyBytes);
+                Assert.NotNull(wrappedKey);
+                Assert.Equal(128, wrappedKey.Length);
+                Assert.False(Utility.AreEqual(keyBytes, wrappedKey), "keyBytes == wrappedKey");
+
+                var unwrappedKey = provider.UnwrapKey(wrappedKey);
+                Assert.NotNull(unwrappedKey);
+                Assert.True(Utility.AreEqual(keyBytes, unwrappedKey), "keyBytes != unwrappedKey");
```
Comment (GeoK) [inline test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultKeyWrapProviderTests.cs]: Consider using _IdentityComparer_ (_AreBytesEqual()_ for this specific case) so you can utilize CompareContext.
There are some other places where you could utilize the context.

We just put out a [Wiki page](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/How-We-Test) w.r.t. testing

File: test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultSecurityKeyConfidentialClientTheoryData.cs (hunk 5 above)
Comment (CBaud) [reply test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultSecurityKeyConfidentialClientTheoryData.cs]: Nope. They both provide the same result. This implementation is simply using string interpolation in order to specify the formatting. "D" is the default behavior when you call .ToString() and could also be explicitly used by calling .ToString("D").

File: test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultKeyWrapProviderTests.cs (hunk 6 above)
Comment (CBaud) [reply test/Microsoft.IdentityModel.Tokens.Extensions.Tests/KeyVaultKeyWrapProviderTests.cs]: Fixed.

**Pull request merged**